src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala

```diff
@@ -0,0 +1,144 @@
+package Cqrs.DbAdapters
+
+import Cqrs.Aggregate
+import Cqrs.Database._
+
+import cats.data.Xor
+import cats.Monad
+import cats.arrow.NaturalTransformation
+import cats._
+import cats.free.Free
+import cats.state._
+import cats.free.Free.{ pure, liftF }
+
+import cats.std.all._
+import lib.foldM
+
+import scala.collection.immutable.TreeMap
+
+object InMemoryDb {
+
+  import Aggregate._
+
+  final case class DbBackend(
+    data: TreeMap[String, TreeMap[String, TreeMap[Int, List[String]]]], // tag -> aggregate id -> version -> event data
+    log: TreeMap[Int, (String, String, Int)],                           // operation nr -> tag, aggregate id, aggregate version
+    lastOperationNr: Int
+  )
+  type Db[A] = State[DbBackend, A]
+
+  def newInMemoryDb: DbBackend = DbBackend(TreeMap.empty, TreeMap.empty, 0)
+
+  def readFromDb[E: EventSerialisation](database: DbBackend, tag: Tag, id: AggregateId, fromVersion: Int): Error Xor List[VersionedEvents[E]] = {
+
+    def getById(id: AggregateId)(t: TreeMap[String, TreeMap[Int, List[String]]]) = t.get(id.v)
+    def unserialise(d: String) = implicitly[EventSerialisation[E]].decode(d)
+    def unserialiseEvents(d: List[String])(implicit t: Traverse[List]): Error Xor List[E] = t.sequence[Xor[Error, ?], E](d map unserialise)
+    def dbRecordToMaybeVersionedEvent(dbrec: (Int, List[String])) = {
```

Comment:
It is unclear to me what the intent of the function is. 

There are two hard things here:

1. The name doesn't describe it for me.
2. The (tuples) types don't describe it, it would be good to use case classes here ( but you would need to come up with good names again :) ).